uber/automated_emails.py

```diff
@@ -12,34 +12,69 @@ class AutomatedEmail:
         Attendee: lambda session: session.staffers(only_staffing=False),
         Group: lambda session: session.query(Group).options(subqueryload(Group.attendees))
     }
-    extra_models = queries  # we've renamed "extra_models" to "queries" but are temporarily keeping the old name for backwards-compatibility
 
-    def __init__(self, model, subject, template, filter, *, sender=None, extra_data=None, cc=None, bcc=None, post_con=False, needs_approval=True):
+    def __init__(self, model, subject, template, filter, *, date_filters=[], sender=None, extra_data=None, cc=None, bcc=None, post_con=False, needs_approval=True):
         self.model, self.template, self.needs_approval = model, template, needs_approval
         self.subject = subject.format(EVENT_NAME=c.EVENT_NAME)
         self.cc = cc or []
         self.bcc = bcc or []
         self.extra_data = extra_data or {}
         self.sender = sender or c.REGDESK_EMAIL
         self.instances[self.subject] = self
+        self.date_filters = date_filters
+
+        # after each daemon run, this is set to the number of emails that would have been sent out but weren't because
+        # they were not marked as approved.
+        self.count_emails_not_sent_need_approval = 0
+
+        # old filter
         if post_con:
             self.filter = lambda x: c.POST_CON and filter(x)
         else:
             self.filter = lambda x: not c.POST_CON and filter(x)
 
+    def filters_run(self, x):
+        if self.filter and not self.filter(x):
+            return False
+
+        for date_filter in listify(self.date_filters):
+            if not date_filter():
+                return False
+
+        return True
+
     def __repr__(self):
         return '<{}: {!r}>'.format(self.__class__.__name__, self.subject)
 
-    def prev(self, x, all_sent=None):
-        if all_sent:
-            return (x.__class__.__name__, x.id, self.subject) in all_sent
+    def already_sent(self, x, previously_sent_emails=None):
+        """
+        Returns true if we have a record of previously sending this email
+
+        Speed Optimization: when using this function as part of batch processing, you can pass in a list of all
+        previously sent emails, previously_sent_emails, in order to avoid having to query the DB for this specific email
+        """
+        if previously_sent_emails:
+            return (x.__class__.__name__, x.id, self.subject) in previously_sent_emails
         else:
             with Session() as session:
                 return session.query(Email).filter_by(model=x.__class__.__name__, fk_id=x.id, subject=self.subject).first()
 
-    def should_send(self, x, all_sent=None):
+    def should_send(self, model_inst, approved_subjects, previously_sent_emails=None):
```

Comment:
FWIW, it would be useful if ``approved_subjects`` defaulted to ``None`` and had this function fill it in if it was missing.  This is useless for the code in this file, but helpful when debugging in a REPL, since it's nice to be able to say ``payment_reminder.should_send(some_attendee)``.

That's the reason why ``previously_send_emails`` defaults to ``None``, though I wouldn't expect anyone to have realized that since I never wrote docstrings for this class :(

Not a huge deal either way though.